aten/src/ATen/SparseGCSTensorImpl.cpp

```diff
@@ -0,0 +1,133 @@
+
+namespace at {
+namespace {
+  DeviceType sparseGCSTensorSetToDeviceType(DispatchKeySet key_set) {
+    if (key_set.has(DispatchKey::SparseGCS_CPU)) {
+      return kCPU;
+    } else if (key_set.has(DispatchKey::SparseGCS_CUDA)) {
+      return kCUDA;
+    } else {
+      AT_ERROR("Cannot construct SparseTensor with non-sparse tensor type ID ", key_set);
+    }
+  }
+}
+
+SparseGCSTensorImpl::SparseGCSTensorImpl(at::DispatchKeySet key_set,
+                                         const caffe2::TypeMeta& data_type)
+  :   SparseGCSTensorImpl(key_set, data_type
+      , at::empty({0}, at::initialTensorOptions().device(sparseGCSTensorSetToDeviceType(key_set)).dtype(ScalarType::Int)) // pointers
+      // indices in case of GCS tensor is always a 1D array so need to init size as {1,0}.
+      , at::empty({0}, at::initialTensorOptions().device(sparseGCSTensorSetToDeviceType(key_set)).dtype(ScalarType::Int)) // indices
+      , at::empty({0}, at::initialTensorOptions().device(sparseGCSTensorSetToDeviceType(key_set)).dtype(data_type)) // values
+      , at::empty({0}, at::initialTensorOptions().device(sparseGCSTensorSetToDeviceType(key_set)).dtype(ScalarType::Int)) // reduction
+) {}
+
+SparseGCSTensorImpl::SparseGCSTensorImpl(at::DispatchKeySet key_set,
+                                         const caffe2::TypeMeta& data_type,
+                                         at::Tensor pointers, at::Tensor indices, at::Tensor values,
+                                         at::Tensor reduction)
+  : TensorImpl(key_set, data_type, values.device()),
+    pointers_(std::move(pointers)),
+    indices_(std::move(indices)),
+    values_(std::move(values)),
+    reduction_(std::move(reduction)) {}
+
+void SparseGCSTensorImpl::resize_and_clear_(int64_t nnz_size, int64_t ptr_size, int64_t redux_size, IntArrayRef size) {
+  // TODO: perform error checking.
+  TORCH_CHECK(size.size() + 1 == redux_size, "size of the reduction array has to be len(sparse.shape)+1, but got: ", redux_size);
+
+  // call pointers().options() here since the struct contructor calls the tensor constructor
+  // with args for device specific init.
+  auto empty_pointers = at::empty(ptr_size, pointers().options());
+  auto empty_indices = at::empty(nnz_size, indices().options());
+  auto empty_values = at::empty(nnz_size, values().options());
+  auto empty_reduction = at::empty(redux_size, reduction().options());
+
+  TORCH_CHECK(empty_indices.scalar_type() == kInt,   "empty_indices must be an int32 type, but got: ", empty_indices.dtype());
+  TORCH_CHECK(empty_pointers.scalar_type() == kInt,  "empty_pointers must be int32 type, but got: ",   empty_pointers.dtype());
+  TORCH_CHECK(empty_reduction.scalar_type() == kInt, "empty_reduction must be int32 type, but got: ",  empty_reduction.dtype());
+
+  // directly set to the member variables. there should be lots of error checking here.
+  pointers_ = empty_pointers;
+  indices_ = empty_indices;
+  values_ = empty_values;
```

Comment:
setting values to empty makes sense, but clearing pointers and indices to have garbage seems super questionable

Reply:
I see this is copied from SparseTensorImpl; I'd encourage you NOT to copy this, and see if you can work out a more sensical way to model this in the new GCS kernels you'll be writing.